Migrated Internationalization documentation to null safety

miquelbeltran opened the pull request:
This was a tough one :) 

In this PR I have migrated the internationalization.md to null safety, as the projects that this document uses were already migrated the work has been mostly using code-excerpts to extract the right lines of code.

In some cases I had to add extra code to show different examples. One section had to be reworded a bit to show the example.
Mostly all the code is extracted in some way.

There is one part in which the doc asks the reader to switch the device to Arabic to see the locale at work, but there was no demo built showing that.

After chatting with @RedBrogdon about this, we decided to remove the mention to Arabic and instead mention that you can switch the device language to Spanish to check that the app text changes, as Spanish is already provided in the sample code.

@sfshaza2

Live version: https://mb-flutter-dev-1.firebaseapp.com/docs/development/accessibility-and-localization/internationalization

sfshaza2 reviewed (commented):
@shihaohong, @miquelbeltran has updated this page to null safety. It LGTM, but can you cast an eye over this?

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -636,11 +630,6 @@ class _NnMaterialLocalizationsDelegate extends LocalizationsDelegate<MaterialLoc
         // for 'en_US' instead.
         decimalFormat: intl.NumberFormat('#,##0.###', 'en_US'),
         twoDigitZeroPaddedFormat: intl.NumberFormat('00', 'en_US'),
-        //...
```
Same question about removing the closing brackets

miquelbeltran replied:
I've been trying to improve this a little bit more but it's very complicated to extract the section of code from the `_NnMaterialLocalizationsDelegate`, so I thought also that it would be beneficial to show the whole class instead instead of doing a very complex RegEx replace. What do you think?

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -619,11 +611,13 @@ not share the same flexibility for number formatting, the formatting
 `_NnMaterialLocalizationsDelegate`:
 
-<!-- skip -->
+<?code-excerpt "add_language/lib/nn_intl.dart (Delegate)" replace="/@(.|\n)*';\n\n//g;/    final (.|\n)*\);\n\n//g"?>
```
Could you add a comment about what the `replace` here does?

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -135,29 +132,32 @@ following instructions to add localized text to your application.
 
 1. Add the `intl` package to the `pubspec.yaml` file:
 
+   <?code-excerpt "gen_l10n_example/pubspec.yaml (Intl)" replace="/(?<!0) # Add this line//g" ?>
```
Add a comment on what this `replace` does

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -69,31 +72,25 @@ dependencies:
-    // ... other locales the app supports
+    const Locale('es', ''), // Spanish, no country code
   ],
-  // ...
-)
```
Should the closing brackets be removed?

miquelbeltran replied:
I changed the script to extract the whole MaterialApp class to avoid confusion.

shihaohong reviewed (commented):
The NNBD migration parts of the code look fine. 

I'm just unsure about some of the code snippet sections that were removed. I'm also not sure what `#docregion` does. One part was also modified incorrectly (ie `MaterialApp` became `Material` in one of the doc samples.

miquelbeltran commented on null_safety_examples/internationalization/add_language/lib/main.dart:
```diff
@@ -5,6 +5,7 @@ import 'nn_intl.dart';
 
 void main() {
   runApp(
+    // #docregion MaterialApp
```
It is part of the code-excerpt logic to copy sections of code into markdown documents. This allows to keep the documents updated with code that stays on example projects. docregion defines where a snippet starts and end. You can learn more about it here: https://github.com/chalin/code_excerpt_updater

miquelbeltran commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -653,19 +642,18 @@ Here's some code that sets the app's language to Nynorsk and
 `localizationsDelegates` list:
 
-<!-- skip -->
+<?code-excerpt "add_language/lib/main.dart (MaterialApp)"?>
```
When the `./tool/refresh-code-excerpts.sh` script runs, it copies over the section of code marked between `docregion` and `enddocregion` named `MaterialApp` from the `main.dart`.

shihaohong replied:
Ahh got it, that makes sense. Thanks!

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
````diff
@@ -69,31 +72,25 @@ dependencies:
-// ...
+<?code-excerpt "gen_l10n_example/lib/main.dart (LocalizationDelegates)" replace="/App.*//g"?>
+```dart
+return Material
````
I wondered if the `replace` fixed this, but it shows in https://mb-flutter-dev-1.firebaseapp.com/docs/development/accessibility-and-localization/internationalization

miquelbeltran replied:
Was an error on my replace RegEx, I have simplified the code-excerpt a bit now.

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
````diff
@@ -69,31 +72,25 @@ dependencies:
+<?code-excerpt "gen_l10n_example/lib/main.dart (LocalizationDelegates)" replace="/App.*//g"?>
+```dart
+return Material
+  title: 'Localizations Sample
````
Looks like the title is cut off as well: 
```suggestion
    title: 'Localizations Sample App',
```

miquelbeltran replied:
fixed now on the latest commit

shihaohong commented on src/docs/development/accessibility-and-localization/internationalization.md:
```diff
@@ -579,38 +572,37 @@ like `r'About $applicationName'`,
     'e.Kr.',
   ],
-  // ...
-}
```
If someone copy/pastes this, it'll result in an error since it's an incomplete map. Is there a way we can preserve the correctness of the objects here and elsewhere? I feel like this will likely be a source of friction for our users.

miquelbeltran commented:
Thanks for the review @shihaohong ! I have addressed your points. Also changed the way the `_NnMaterialLocalizationsDelegate` is extracted into the markdown document, let me know what you think :)

shihaohong reviewed (approved):
LGTM! I think leaving in the full sample despite it making it more verbose is better than leaving it short and perhaps confusing. Thanks for the PR :)

sfshaza2 reviewed (approved):
A huge thanks to you both, @miquelbeltran and @shihaohong!
